forte/data/data_store.py

```diff
@@ -1044,20 +1072,166 @@ def _create_new_entry(
         tid: int = self._new_tid() if tid is None else tid
         entry: List[Any] = []
 
-        entry.extend(attribute_data)
-
         entry += [tid, type_name]
         entry += self._default_attributes_for_type(type_name)
 
         return entry
 
+    def get_datastore_attr_idx(self, type_name: str, attr: str) -> int:
+        r"""This function returns the index of where a given attribute
+        `attr` is stored in the Data Store entry of type `type_name`
+
+        Args:
+            type_name (str): The fully qualified type name of the entry.
+            attr (str): The name of the attribute whose index needs to be fetched.
+
+        Returns:
+            An integer representing the attributes position in the Data
+            Store entry.
+        """
+        try:
+            return DataStore._type_attributes[type_name][
+                constants.ATTR_INFO_KEY
+            ][attr][constants.ATTR_INDEX_KEY]
+        except KeyError as e:
+            raise KeyError(
+                f"Attribute {attr} is not a part "
+                f"of type attributes for {type_name}"
+                f"or an entry of type {type_name} does not exist"
+            ) from e
+
+    def initialize_and_validate_entry(
+        self, entry: List, attribute_data: List
+    ) -> List:
+        r"""This function performs validation checks on the initial
+        attributes added to a data store entry. This functions also
+        modifies the value of certain attributes to fit data store's
+        purpose of storing primitive types. For example, In the data store
+        entry of type :class:`~forte.data.ontology.top.Group`, attribute
+        `member_type` is converted from an object to `str`. When initializing
```

Comment:
Sometimes we probably support initializing attributes with pure string in the first place. Would this be supported?

Reply:
No if a pure string is passed into `get_full_module name`, its returns its value as `str`. A solution can be that if a string is passed, we can validate it by first converting it to an object using `get_class` and raise an error if its not a valid entry. Does that sound okay?

Reply:
We need to look at the path when creating an entry now. At DataPack level, we should support creating an entry object via "class" or "classname". Note that in datastore we can check whether a string is a an entry using the _type_attribute_dict.

And we shouldn't convert back and forth between these two types, this defeats the purpose of being "data efficient", we should support string and class naturally.

Reply:
The reason I mentioned using `get_class` was to validate if the string passed is a path to a valid class. 

Also, with regards to `member_type`, are you saying that during initializing a data store entry, it can be a string rather than an object?

Reply:
So we don't want to have calls to validate classes during entry creation. This might be slow. Recall that the reason to include a datastore is to reduce class creation overhead. We can't afford new overhead during entry creation.

It is fine to collect the classes and types needed as a one-time process, but we cannot have heavy procedures during `initialize_and_validate_entry`.

You can also notice that in data_store most of the data are stored in primitive forms such as strings, like 
https://github.com/asyml/forte/blob/1fcaebcee6e7ab0ea4e4b8defc8c9192c235d6e5/forte/data/data_store.py#L982

Reply:
So based on what I understand, there's a need to reduce the overhead introduced by the calls of `self._is_subclass` because it tried to create classes. In that case we can just get rid of this check and just directly store the output of `get_full_module_name` for attributes that are stored in the data store entry as strings.

Reply:
1. `_is_subclass` may not create classes, it could use an internal cache. It is supposed to be an efficient implementation. It only checks a string containment in a small dict when the setup is correct.
2. `_is_subclass` is not a tool to be used for general class checking. It is only designed for checking subclass among Forte entries.
3. where do we need to call `get_full_module_name`? We also don't want to call this function if it introduced overhead
4. Instead of `get rid of the check`, the desired approach here is to understand what's being passed in and try to validate this in a quick way. 

I don't think we want to do [this](https://github.com/asyml/forte/blob/1fcaebcee6e7ab0ea4e4b8defc8c9192c235d6e5/forte/data/data_store.py#L1200) ` issubclass(get_class(type_name), cls) for cls in (Link, MultiPackLink)`, this involves a lot of class manipulation. Actually, why not `self._is_subclass(type_name, cls) for cls in (Link, MultiPackLink)`?

We need to have a careful review of the efficiency here, minimizing conversion between class and string etc. @mylibrar could you help monitoring this?

Reply:
1, 2: I understand and I will use it accordingly
3: So [here](https://github.com/Pushkar-Bhuse/forte/blob/d8cab15cef6fce6bbcf6ef580200a179e922f503/forte/data/data_store.py#L1201), `attribute_data[0]` stores the value of `self.member_type` which is a `class`. Thus, before storing it in data store, we need to convert it into a string using `get_full_module_name`. This was something we did before as well [here](https://github.com/asyml/forte/blob/e86602f55530a739a8e9d165c8cb51d564d88474/forte/data/entry_converter.py#L90)
4. The reason I feel we can get rid of the check is that this check validates if the `member_type` is a valid `Entry`. But since this value is derived from [here](https://github.com/Pushkar-Bhuse/forte/blob/d8cab15cef6fce6bbcf6ef580200a179e922f503/forte/data/ontology/top.py#L328), it will always be a valid `Entry`

For your last point, you're right, I will update that.